File: src/grpcbigbuffer/block_builder.py
import json
import os.path
import warnings
from hashlib import sha3_256
from io import BufferedReader
from itertools import zip_longest
from typing import Any, List, Dict, Union, Tuple
from grpcbigbuffer import buffer_pb2
from google.protobuf.message import Message, DecodeError
from google._upb._message import RepeatedCompositeContainer

from grpcbigbuffer.client import generate_random_dir, block_exists, move_to_block_dir, copy_to_block_dir, \
    get_hash_from_block
from grpcbigbuffer.utils import Enviroment, CHUNK_SIZE, METADATA_FILE_NAME, WITHOUT_BLOCK_POINTERS_FILE_NAME, \
    get_file_hash, create_lengths_tree, encode_bytes


def is_block(bytes_obj: bytes, blocks: List[bytes]) -> bool:
    try:
        block = buffer_pb2.Buffer.Block()
        with warnings.catch_warnings():
            warnings.simplefilter("ignore", category=RuntimeWarning)
            block.ParseFromString(bytes_obj)
        return get_hash_from_block(block, internal_block=False, hexadecimal=False) in blocks
    except DecodeError:
        pass
    return False


def get_position_length(varint_pos: int, buffer: bytes) -> int:
    """
    Returns the value of the varint at the given position in the Protobuf buffer.
    """
    value = 0
    shift = 0
    index = varint_pos
    while True:
        byte = buffer[index]
        value |= (byte & 0x7F) << shift
        if (byte & 0x80) == 0:
            break
        shift += 7
        index += 1
    return value


def get_hash(block: buffer_pb2.Buffer.Block) -> str:
    for _hash in block.hashes:
        if _hash.type == Enviroment.hash_type:
            return _hash.value.hex()
    raise Exception('gRPCbb: any hash of type ' + Enviroment.hash_type.hex())


def get_block_length(block_id: str) -> int:
    if os.path.isfile(Enviroment.block_dir + block_id):
        return os.path.getsize(Enviroment.block_dir + block_id)
    elif os.path.isdir(Enviroment.block_dir + block_id):
        raise Exception('gRPCbb: error on compute_real_lengths, multiblock blocks dont supported.'
                        + Enviroment.block_dir + block_id)
    else:
        raise Exception('gRPCbb: error on compute_real_lengths, block does not in block registry. '
                        + Enviroment.block_dir + block_id)


def search_on_message_real(
        message: Message,
        pointers: List[int],
        initial_position: int,
        real_initial_position: int,
        blocks: List[bytes],
        container: List[Tuple[str, List[int]]],
        real_lengths: Dict[int, Tuple[int, int, bool]],
):
    position: int = initial_position
    real_position: int = real_initial_position
    for field, value in message.ListFields():
        if isinstance(value, RepeatedCompositeContainer):
            for element in value:
                position += 1
                if position not in real_lengths.keys():
                    position += len(encode_bytes(element.ByteSize())) + element.ByteSize()
                    real_position += 1 + len(encode_bytes(element.ByteSize())) + element.ByteSize()
                    continue
                try:
                    message_size = real_lengths[position][0]
                except KeyError:
                    raise Exception(
                        'gRPCbb block builder error, real lengths not in ' + str(position) + '. ' + str(real_lengths))
                search_on_message_real(
                    message=element,
                    pointers=pointers + [real_position + 1],
                    initial_position=position + len(encode_bytes(element.ByteSize())),
                    real_initial_position=real_position + 1 + len(encode_bytes(message_size)),
                    blocks=blocks,
                    container=container,
                    real_lengths=real_lengths
                )
                position += len(encode_bytes(element.ByteSize())) + element.ByteSize()
                real_position += 1 + len(encode_bytes(message_size)) + message_size

        elif isinstance(value, Message):
            position += 1
            if position not in real_lengths.keys():
                position += len(encode_bytes(value.ByteSize())) + value.ByteSize()
                real_position += 1 + len(encode_bytes(value.ByteSize())) + value.ByteSize()
                continue
            try:
                message_size = real_lengths[position][0]
            except KeyError:
                raise Exception(
                    'gRPCbb block builder error, real lengths not in ' + str(position) + '. ' + str(real_lengths))
            search_on_message_real(
                message=value,
                pointers=pointers + [real_position + 1],
                initial_position=position + len(encode_bytes(value.ByteSize())),
                real_initial_position=real_position + 1 + len(encode_bytes(message_size)),
                blocks=blocks,
                container=container,
                real_lengths=real_lengths
            )
            position += len(encode_bytes(value.ByteSize())) + value.ByteSize()
            real_position += 1 + len(encode_bytes(message_size)) + message_size

        elif type(value) == bytes and is_block(value, blocks):
            block = buffer_pb2.Buffer.Block()
            block.ParseFromString(value)
            container.append((get_hash(block), pointers + [real_position + 1]))
            block_length: int = get_block_length(get_hash(block))
            position += 1
            try:
                if (real_lengths[position][0] != block_length):
                    raise Exception('Error on gRPCbb: block_builder method computing real message positions. ',
                                    position, real_lengths[position], block_length)
            except KeyError:
                raise Exception(
                    'gRPCbb block builder error, real lengths not in ' + str(position) + '. ' + str(real_lengths))
            position += len(encode_bytes(block.ByteSize())) + block.ByteSize()
            real_position += 1 + len(encode_bytes(block_length)) + block_length

        elif type(value) == bytes or type(value) == str:
            position += 1 + len(encode_bytes(len(value))) + len(value)
            real_position += 1 + len(encode_bytes(len(value))) + len(value)

        else:
            try:
                temp_message = type(message)()
                temp_message.CopyFrom(message)
                for field_name, _ in temp_message.ListFields():
                    if field_name.index != field.index:
                        temp_message.ClearField(field_name.name)
                position += temp_message.ByteSize()
                real_position += temp_message.ByteSize()
            except Exception as e:
                raise Exception('gRPCbb block builder error obtaining the length of a primitive value :' + str(e))


def search_on_message(
        message: Message,
        pointers: List[int],
        initial_position: int,
        blocks: List[bytes],
        container: Dict[str, List[List[int]]]
):
    """
       Search_on_message makes a tree search of the protobuf object (attr. message) and stores all the buffer block
        identifier instances with its indexes (ascendant order) on the container dictionary.
        It allows to know where the buffer needs to be changed when the buffer block substitute the block identifier.
       """
    position: int = initial_position
    for field, value in message.ListFields():
        if isinstance(value, RepeatedCompositeContainer):
            for element in value:
                search_on_message(
                    message=element,
                    pointers=pointers + [position + 1],
                    initial_position=position + 1 + len(encode_bytes(element.ByteSize())),
                    blocks=blocks,
                    container=container
                )
                position += 1 + len(encode_bytes(element.ByteSize())) + element.ByteSize()

        elif isinstance(value, Message):
            search_on_message(
                message=value,
                pointers=pointers + [position + 1],
                initial_position=position + 1 + len(encode_bytes(value.ByteSize())),
                blocks=blocks,
                container=container
            )
            position += 1 + len(encode_bytes(value.ByteSize())) + value.ByteSize()

        elif type(value) == bytes and is_block(value, blocks):
            block = buffer_pb2.Buffer.Block()
            block.ParseFromString(value)

            _block_hash = get_hash(block)
            _list_of_pointers = pointers + [position + 1]
            if _block_hash not in container:
                container[_block_hash] = [_list_of_pointers]
            else:
                container[_block_hash].append(_list_of_pointers)

            position += 1 + len(encode_bytes(block.ByteSize())) + block.ByteSize()

        elif type(value) == bytes or type(value) == str:
            position += 1 + len(encode_bytes(len(value))) + len(value)

        else:
            try:
                temp_message = type(message)()
                temp_message.CopyFrom(message)
                for field_name, _ in temp_message.ListFields():
                    if field_name.index != field.index:
                        temp_message.ClearField(field_name.name)
                position += temp_message.ByteSize()
            except Exception as e:
                raise Exception('gRPCbb block builder error obtaining the length of a primitive value :' + str(e))


def compute_real_lengths(tree: Dict[int, Union[Dict, str]], buffer: bytes) -> Dict[int, Tuple[int, int, bool]]:
    """
    Given the pointer's tree with block id's as the leafs it will return a dict of pointers with its
    real length, compressed format (of the input buffer) length, and a boolean saying if it's the pointer
    of a block id message or not (tree's leaf or not).

    :param tree:Tree of pointers as nodes and block id's as leafs.
    :type tree: Dict[int, Union[Dict, str]]
    :param buffer:Buffer of the compressed object.
    :type buffer: bytes
    :return: A dict of pointers with its real and compressed lengths and if it's leaf or not.
    :rtype: Dict[int, Tuple[int, int, bool]]
    """
    def traverse_tree(internal_tree: Dict, internal_buffer: bytes, initial_total_length: int) \
            -> Tuple[int, Dict[int, Tuple[int, int, bool]]]:

        real_lengths: Dict[int, Tuple[int, int, bool]] = {}
        total_tree_length: int = 0
        total_block_length: int = 0
        for key, value in internal_tree.items():
            if isinstance(value, dict):
                initial_length: int = get_position_length(key, internal_buffer)
                real_length, internal_lengths = traverse_tree(
                    value, internal_buffer, initial_length
                )
                real_lengths[key] = (real_length, initial_length, False)
                real_lengths.update(internal_lengths)
                total_tree_length += real_length + len(encode_bytes(real_length)) + 1

                block_length: int = initial_length + len(encode_bytes(initial_length)) + 1
                total_block_length += block_length

            else:
                b = buffer_pb2.Buffer.Block()
                h = buffer_pb2.Buffer.Block.Hash()
                h.type = Enviroment.hash_type
                h.value = bytes.fromhex(value)
                b.hashes.append(h)
                b_length: int = len(b.SerializeToString())

                real_length: int = get_block_length(value)
                real_lengths[key] = (real_length, b_length, True)
                total_tree_length += real_length + len(encode_bytes(real_length)) + 1

                block_length: int = b_length + len(encode_bytes(b_length)) + 1
                total_block_length += block_length

        if initial_total_length < total_block_length:
            raise Exception('Error on compute real lengths, block length cant be greater than the total length',
                            initial_total_length, total_block_length)

        total_tree_length += initial_total_length - total_block_length

        return total_tree_length, real_lengths

    #  For the case when are duplicate blocks, the lengths tree needs to be sorted.
    return dict(sorted(traverse_tree(tree, buffer, len(buffer))[1].items()))


def generate_buffer(buffer: bytes, lengths: Dict[int, Tuple[int, int, bool]]) -> List[bytes]:
    """
    Iterates over the buffer, replacing the compressed buffer with the real buffer,
    replacing the compressed lengths of each pointer with its real length.
    It is returned in list format since it is not necessary to return the entire buffer,
    the content of the blocks does not need to be loaded into memory.

    :param buffer: Compressed buffer
    :type buffer: bytes
    :param lengths:  A dict of pointers with its real and compressed lengths and if it's leaf or not.
    :type lengths: Dict[int, Tuple[int, int, bool]]
    :return: The inter-block buffers list with the real lengths.
    :rtype: List[bytes]
    """
    list_of_bytes: List[bytes] = []
    new_buff: bytes = b''
    i: int = 0
    for key, value in lengths.items():
        new_buff += buffer[i:key] + encode_bytes(value[0])
        i = key + len(encode_bytes(value[1]))
        if value[2]:
            i += value[1]
            list_of_bytes.append(new_buff)
            new_buff = b''

    return list_of_bytes + [buffer[i:]]


def generate_id(buffers: List[bytes], blocks: List[bytes]) -> bytes:
    hash_id = sha3_256()
    for buffer, block in zip_longest(buffers, blocks):
        if buffer:
            hash_id.update(buffer)
        if block:
            with BufferedReader(open(Enviroment.block_dir + block.hex(), 'rb')) as f:
                while True:
                    f.flush()
                    piece: bytes = f.read(CHUNK_SIZE)
                    if len(piece) == 0:
                        break
                    hash_id.update(piece)
    return hash_id.digest()


def purify_buffer(buff: bytes) -> bytes:
    return buff


def build_multiblock(
        pf_object_with_block_pointers: Any,
        blocks: List[bytes]
) -> Tuple[bytes, str]:
    container: Dict[str, List[List[int]]] = {}
    search_on_message(
        message=pf_object_with_block_pointers,
        pointers=[],
        initial_position=0,
        blocks=blocks,
        container=container
    )

    tree: Dict[int, Union[Dict, str]] = create_lengths_tree(
        pointer_container=container
    )

    # no importa para duplicidad
    real_lengths: Dict[int, Tuple[int, int, bool]] = compute_real_lengths(
        tree=tree,
        buffer=pf_object_with_block_pointers.SerializeToString()
    )

    # no importa para duplicidad
    new_buff: List[bytes] = generate_buffer(
        buffer=pf_object_with_block_pointers.SerializeToString(),
        lengths=real_lengths
    )

    object_id: bytes = generate_id(
        buffers=new_buff,
        blocks=blocks
    )
    cache_dir: str = generate_random_dir() + '/'
    _json: List[Union[
        int,
        Tuple[str, List[int]]
    ]] = []

    container_real_lengths: List[Tuple[str, List[int]]] = []
    search_on_message_real(
        message=pf_object_with_block_pointers,
        pointers=[],
        initial_position=0,
        real_initial_position=0,
        blocks=blocks,
        container=container_real_lengths,
        real_lengths=real_lengths
    )

    for i, (b1, b2) in enumerate(zip_longest(new_buff, container_real_lengths)):
        _json.append(i + 1)
        with open(cache_dir + str(i + 1), 'wb') as f:
            f.write(b1)

        if b2:
            _json.append((b2[0], b2[1]))

    with open(cache_dir + METADATA_FILE_NAME, 'w') as f:
        json.dump(_json, f)

    with open(cache_dir + WITHOUT_BLOCK_POINTERS_FILE_NAME, 'wb') as f:
        f.write(pf_object_with_block_pointers.SerializeToString())

    return object_id, cache_dir


def create_block(file_path: str, copy: bool = False) -> Tuple[bytes, buffer_pb2.Buffer.Block]:
    file_hash: str = get_file_hash(file_path=file_path)
    if not block_exists(block_id=file_hash):
        if copy and not copy_to_block_dir(
                file_hash=file_hash,
                file_path=file_path
        ) or \
                not copy and not move_to_block_dir(
            file_hash=file_hash,
            file_path=file_path
        ):
            raise Exception('gRPCbb error creating block, file could not be moved.')

    file_hash: bytes = bytes.fromhex(file_hash)

    block = buffer_pb2.Buffer.Block()
    h = buffer_pb2.Buffer.Block.Hash()
    h.type = Enviroment.hash_type
    h.value = file_hash
    block.hashes.append(h)

    return file_hash, block

File: pyo3_generator_test/src/block_builder.rs
// This is a simplified translation of the Python code into Rust, focusing on the main structure and logic.
// It is not a direct line-by-line translation due to differences between Python and Rust.
// Some Python-specific constructs and libraries (e.g., gRPC, protobuf) do not have direct equivalents in Rust,
// so this translation assumes the existence of similar functionality in Rust libraries or crates.

use std::collections::HashMap;
use std::fs::File;
use std::io::{BufReader, Read};
use std::path::Path;

use prost::Message;
use sha3::{Digest, Sha3_256};

// Assuming equivalent functionality is provided by a Rust gRPC library or crate.
use pee_rpc::{
    buffer_proto, client, utils, BufferBlock, Environment, CHUNK_SIZE, METADATA_FILE_NAME,
    WITHOUT_BLOCK_POINTERS_FILE_NAME,
};

// Rust structs and enums to represent the necessary types and logic.
// ...

// Functions translated from Python to Rust.

fn is_block(bytes_obj: &[u8], blocks: &[Vec<u8>]) -> bool {
    // Implementation of is_block in Rust.
    // ...
}

fn get_position_length(varint_pos: usize, buffer: &[u8]) -> usize {
    // Implementation of get_position_length in Rust.
    // ...
}

fn get_hash(block: &BufferBlock) -> String {
    // Implementation of get_hash in Rust.
    // ...
}

fn get_block_length(block_id: &str) -> usize {
    // Implementation of get_block_length in Rust.
    // ...
}

fn search_on_message_real(
    // Parameters converted to equivalent Rust types.
    // ...
) {
    // Implementation of search_on_message_real in Rust.
    // ...
}

fn search_on_message(
    // Parameters converted to equivalent Rust types.
    // ...
) {
    // Implementation of search_on_message in Rust.
    // ...
}

fn compute_real_lengths(
    // Parameters converted to equivalent Rust types.
    // ...
) -> HashMap<usize, (usize, usize, bool)> {
    // Implementation of compute_real_lengths in Rust.
    // ...
}

fn encode_bytes(mut n: u64) -> Vec<u8> {
    let mut buf = Vec::new();
    loop {
        let mut towrite = (n & 0x7f) as u8;
        n >>= 7;
        if n != 0 {
            towrite |= 0x80;
        }
        buf.push(towrite);
        if n == 0 {
            break;
        }
    }
    buf
}

fn generate_buffer(buffer: &[u8], lengths: &HashMap<usize, (usize, usize, bool)>) -> Vec<Vec<u8>> {
    /*
    Iterates over the buffer, replacing the compressed buffer with the real buffer,
    replacing the compressed lengths of each pointer with its real length.
    It is returned in list format since it is not necessary to return the entire buffer,
    the content of the blocks does not need to be loaded into memory.

    :param buffer: Compressed buffer
    :type buffer: bytes
    :param lengths:  A dict of pointers with its real and compressed lengths and if it's leaf or not.
    :type lengths: Dict[int, Tuple[int, int, bool]]
    :return: The inter-block buffers list with the real lengths.
    :rtype: List[bytes]
    */
    let mut list_of_bytes: Vec<Vec<u8>> = Vec::new();
    let mut new_buff: Vec<u8> = Vec::new();
    let mut i: usize = 0;
    for (&key, &(real_length, compressed_length, is_leaf)) in lengths {
        new_buff.extend(&buffer[i..key]);
        new_buff.extend(encode_bytes(real_length));
        i = key + encode_bytes(compressed_length).len();
        if is_leaf {
            i += compressed_length;
            list_of_bytes.push(new_buff.clone());
            new_buff.clear();
        }
    }
    list_of_bytes.push(buffer[i..].to_vec());
    list_of_bytes
}


// Assuming `Environment` and `CHUNK_SIZE` are defined elsewhere in the code.
// `Environment` should provide access to the `block_dir` path.
struct Environment;

impl Environment {
    fn block_dir() -> &'static str {
        // Return the directory path where blocks are stored.
        // This should be defined based on your application's needs.
        "/path/to/block/dir/"
    }
}

fn generate_id(buffers: &[Vec<u8>], blocks: &[Vec<u8>]) -> io::Result<Vec<u8>> {
    let mut hash_id = Sha3_256::new();
    for (buffer, block) in buffers.iter().zip_longest(blocks.iter()) {
        if let Some(buf) = buffer {
            hash_id.update(buf);
        }
        if let Some(block) = block {
            let block_path = format!("{}{}", Environment::block_dir(), hex::encode(block));
            let mut file = BufReader::new(File::open(block_path)?);
            let mut piece = vec![0u8; CHUNK_SIZE];
            loop {
                let bytes_read = file.read(&mut piece)?;
                if bytes_read == 0 {
                    break;
                }
                hash_id.update(&piece[..bytes_read]);
            }
        }
    }
    Ok(hash_id.finalize().to_vec())
}

fn purify_buffer(buff: &[u8]) -> Vec<u8> {
    buff
}


// The protobuf message type, as well as the specific hash and block types, will need to be defined
// according to the .proto files used in your project. For this example, we will use placeholder types.
struct ProtobufMessageWithBlockPointers; // Placeholder for the actual Protobuf message type.

impl ProtobufMessageWithBlockPointers {
    fn serialize_to_vec(&self) -> Vec<u8> {
        // Serialize the message to a byte vector.
        // This should be implemented using prost or another Protobuf library.
        vec![]
    }
}

fn build_multiblock(
    pf_object_with_block_pointers: ProtobufMessageWithBlockPointers,
    blocks: Vec<Vec<u8>>,
) -> Result<(Vec<u8>, String), Box<dyn std::error::Error>> {
    let mut container = HashMap::new();
    search_on_message(
        &pf_object_with_block_pointers,
        &mut vec![],
        0,
        &blocks,
        &mut container,
    );

    let tree = create_lengths_tree(&container);
    let buffer = pf_object_with_block_pointers.serialize_to_vec();
    let real_lengths = compute_real_lengths(&tree, &buffer);
    let new_buff = generate_buffer(&buffer, &real_lengths);
    let object_id = generate_id(&new_buff, &blocks);
    let cache_dir = generate_random_dir() + "/";
    let mut _json = vec![];

    let mut container_real_lengths = vec![];
    search_on_message_real(
        &pf_object_with_block_pointers,
        &mut vec![],
        0,
        0,
        &blocks,
        &mut container_real_lengths,
        &real_lengths,
    );

    for (i, (b1, b2)) in new_buff.into_iter().zip_longest(container_real_lengths).enumerate() {
        _json.push(i + 1);
        let mut file = File::create(PathBuf::from(&cache_dir).join(i.to_string()))?;
        file.write_all(&b1)?;

        if let Some(b2) = b2 {
            _json.push((b2.0, b2.1));
        }
    }

    let metadata_file_path = PathBuf::from(&cache_dir).join(METADATA_FILE_NAME);
    let mut metadata_file = File::create(metadata_file_path)?;
    metadata_file.write_all(serde_json::to_string(&_json)?.as_bytes())?;

    let without_block_pointers_file_path = PathBuf::from(&cache_dir).join(WITHOUT_BLOCK_POINTERS_FILE_NAME);
    let mut without_block_pointers_file = File::create(without_block_pointers_file_path)?;
    without_block_pointers_file.write_all(&pf_object_with_block_pointers.serialize_to_vec())?;

    Ok((object_id, cache_dir))
}

fn create_block(file_path: &Path, copy: bool) -> io::Result<(Vec<u8>, BufferBlock)> {
    let file_hash_str = get_file_hash(file_path)?;

    if !block_exists(&file_hash_str) {
        if copy {
            copy_to_block_dir(&file_hash_str, file_path)?;
        } else {
            move_to_block_dir(&file_hash_str, file_path)?;
        }
    }

    let file_hash_bytes = hex::decode(file_hash_str).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;

    let mut block = BufferBlock::default();
    block.hashes.push(HashType {
        hash_type: Environment::hash_type(), // Assuming Environment is a struct that provides the hash type.
        value: file_hash_bytes.clone(),
    });

    Ok((file_hash_bytes, block))
}

// Additional Rust-specific code to support the translated functions.
// ...

fn main() {
    // Example usage of the translated functions in a Rust main function.
    // ...
}
